joining/src/main/scala/akka/cluster/bootstrap/dns/HeadlessServiceDnsBootstrap.scala

```diff
@@ -0,0 +1,278 @@
+ * to join itself (becoming the first node of the cluster, that all other nodes will join).
+ *
+ * The decision of joining "self" is made by deterministically sorting the discovered service IPs
+ * and picking the *lowest* address.
+ *
+ * If this node is the one with the lowest address in the deployment, it will join itself and other nodes will notice
+ * this via the contact-point probing mechanism and join this node. Please note while the cluster is "growing"
+ * more nodes become aware of the cluster and start returning the seed-nodes in their contact-points, thus the joining
+ * process becomes somewhat "epidemic". Other nodes may get to know about this cluster by contacting any other node
+ * that has joined it already, and they may join any seed-node that they retrieve using this method, as effectively
+ * this will mean it joins the "right" cluster.
+ *
+ * CAVEATS:
+ * There is a slight timing issue, that may theoretically appear in this bootstrap process.
+ * FIXME explain the races
+ */
+// also known as the "Baron von Bootstrappen"
+@InternalApi
+final class HeadlessServiceDnsBootstrap(settings: ClusterBootstrapSettings)
+    extends Actor
+    with ActorLogging
+    with Timers {
+
+  import HeadlessServiceDnsBootstrap.Protocol._
+  import HeadlessServiceDnsBootstrap._
+  import context.dispatcher
+
+  private val dns = IO(Dns)(context.system)
+  private val cluster = Cluster(context.system)
+
+  private val TimerKeyResolveDNS = "resolve-dns-key"
+
+  private var lastContactsObservation: DnsServiceContactsObservation =
+    DnsServiceContactsObservation(Long.MaxValue, Nil)
+
+  /** Awaiting initial signal to start the bootstrap process */
+  override def receive: Receive = {
+    case InitiateBootstraping ⇒
+      val serviceName = settings.effectiveServiceName(context.system)
+
+      log.info("Locating service members, via DNS lookup: {}", serviceName)
+      resolve(serviceName, settings.dnsResolveTimeout).pipeTo(self)
+
+      context become bootstraping(serviceName, sender())
+  }
+
+  /** In process of searching for seed-nodes */
+  def bootstraping(serviceName: String, replyTo: ActorRef): Receive = {
+    case Internal.AttemptResolve(name, timeout) ⇒
+      resolve(name, timeout).pipeTo(self)
+
+    case DnsResolvedTo(name, contactPoints) ⇒
+      onContactPointsResolved(name, contactPoints)
+
+    case ex: Failure ⇒
+      log.warning("Resolve attempt failed! Cause: {}", ex.cause)
+      scheduleNextResolve(serviceName, settings.dnsResolveTimeout)
+
+    case ObtainedHttpSeedNodesObservation(infoFromAddress, observedSeedNodes) ⇒
+      log.info("Contact point [{}] returned [{}] seed-nodes [{}], initiating cluster joining...", infoFromAddress,
+        observedSeedNodes.size, observedSeedNodes)
+
+      replyTo ! "JOINING SEED NODES!" // FIXME nicer typed message
+
+      val seedNodesList = observedSeedNodes.toList
+      cluster.joinSeedNodes(seedNodesList)
```

Comment:
This is used for joining existing cluster. Remove the selfAddress from the seedNodesList so that there is no chance that it's joining itself.